src/scheduler/uncontrolled_randomness.rs

```diff
@@ -0,0 +1,124 @@
+use crate::runtime::task::TaskId;
+use crate::scheduler::{Schedule, Scheduler};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum ScheduleRecord {
+    Task(Option<TaskId>, Vec<TaskId>, bool),
+    Random(u64),
+}
+
+/// An `UncontrolledRandomnessCheckScheduler` checks whether a given program is deterministic
+/// by wrapping an inner `Scheduler` and, for each schedule generated by that scheduler,
+/// replaying the schedule once. When doing the replay, we check that the schedule is still
+/// valid and that the set of runnable tasks is the same at each step.
+/// Violations of these checks means that the program exhibits randomness which is not
+/// under Shuttle's control.
+#[derive(Debug)]
+pub struct UncontrolledRandomnessCheckScheduler<S: Scheduler> {
+    scheduler: Box<S>,
+    recording: bool,
+    previous_schedule: Vec<ScheduleRecord>,
+    current_step: usize,
+}
+
+impl<S: Scheduler> UncontrolledRandomnessCheckScheduler<S> {
+    /// Create a new `UncontrolledRandomnessCheckScheduler` by wrapping the given `Scheduler` implementation.
+    pub fn new(scheduler: S) -> Self {
+        Self {
+            scheduler: Box::new(scheduler),
+            previous_schedule: Vec::new(),
+            recording: false,
+            current_step: 0,
+        }
+    }
+}
+
+impl<S: Scheduler> Scheduler for UncontrolledRandomnessCheckScheduler<S> {
+    fn new_execution(&mut self) -> Option<Schedule> {
+        // Dummy schedule. We do this instead of doing `self.scheduler.new_execution`
+        // as that would cause the schedule to be ran half as many times as intended.
+        let mut out = Some(Schedule::new(12345));
```

Comment:
Just to be explicit that there's nothing special about this number
```suggestion
        let mut out = Some(Schedule::new(0));
```